src/test/java/api/requests/RequestsAPIRetrievalTests.java

```diff
@@ -687,7 +687,7 @@ void requestExpirationDateShouldStoreInTenantsTimeZone() {
         .by(sponsor)
         .proxiedBy(proxy)
         .fulfilToHoldShelf()
-        .withRequestExpiration(LocalDate.of(2017, 7, 30))
+        .withRequestExpiration(LocalDate.of(2017, 7, 31))
```

Comment:
Why has the date been changed in this test?

Reply:
After double-checking, here is what I am seeing and understanding.

The [RequestFromRepresentationService.getRequestFrom()](https://github.com/folio-org/mod-circulation/blob/440d37bee797dd6ea680f1573168c303117f5a3d/src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java#L76) is calling `Request::from()` which converts the json data into a Request object.

The problem happens when the date string (without time) is converted (via the `r.map()` and `Request::from` calls) into a `DateTime`.
There is no time zone data available at this point and when the date string (without time) is then converted into a a `DateTime`, the UTC date is automatically used.

Then, the immediatly following line fetches the time zone and converts this newly created `DateTime` into the appropriate local time zone.

Any dates without time passed to this therefore must be in UTC.
The `LocalDate.of(2017, 7, 30)` needs to become `LocalDate.of(2017, 7, 31)` to correctly represent UTC date (that will eventually be converted into the local time zone via [Request.truncateRequestExpirationDateToTheEndOfTheDay()](https://github.com/folio-org/mod-circulation/blob/440d37bee797dd6ea680f1573168c303117f5a3d/src/main/java/org/folio/circulation/domain/Request.java#L348).

Reply:
> Any dates without time passed to this therefore must be in UTC.

I don't think I understand this, a date without a time cannot be in any time zone.

Reply:
> The problem happens when the date string (without time) is converted (via the r.map() and Request::from calls) into a DateTime.

I agree that is a challenge (and possible a mistake)

Reply:
> The problem happens when the date string (without time) is converted (via the r.map() and Request::from calls) into a DateTime.

How did you come to the conclusion that the request expiration date does not have a time component, [the JSON schema for a request|https://github.com/folio-org/mod-circulation-storage/blob/7eb6639ee0a1d78e24627c9ccae4de631b3b764d/ramls/request.json#L178] says it is a date and time.

Reply:
```
I don't think I understand this, a date without a time cannot be in any time zone.
```
If only that were true.

Dates, unfortunately, are time zone sensitive, even if there is no provided time.
This is essentially a problem from dates (without time) derived from a date (with time) or vice-versa.

Consider this:
https://www.timeanddate.com/worldclock/converter.html?iso=20201117T005900&p1=1440&p2=179

If I had a Date of `Tue, Nov 17, 2020`, and the date was derived from a UTC date, then the New York date representation would instead be `Mon, Nov 16, 2020`.

Whether "today" is `Mon, Nov 16, 2020` or whether "today" is `Tue, Nov 17, 2020` is dependant on your time zone.

Reply:
> Dates, unfortunately, are time zone sensitive, even if there is no provided time.

I don't understand this. By definition, a date cannot be relative to a time zone, there is insufficient information to determine that.

I can understand that some assumptions can be made about a date only in order to infer what the time and zone should be (e.g. to use the information from the machine the code is running on)

> If I had a Date of Tue, Nov 17, 2020, and the date was derived from a UTC date, then the New York date representation would instead be Mon, Nov 16, 2020.

How can the _date_ be derived from a UTC date, a date cannot be in UTC, only a date and time can be in a zone.

I feel stupid. What I am missing that means this makes sense to you and not to me?

> Whether "today" is Mon, Nov 16, 2020 or whether "today" is Tue, Nov 17, 2020 is dependant on your time zone.

I agree that what the date today is, is based upon where you are in the world. I don't think that can be worked out by a date alone.

Reply:
> The test is then passing a date without the time.

What property is the test providing a date for?

Could this be resolved by providing a date and time if that is what is expected?

Reply:
> This results in an incorrect test because Midnight of 2017, 7, 30 in UTC is not the midnight of 2017, 7, 31 in New York.

Why did this test work / pass with the old code and not the new?

Is this a bug in the production code, a bug in the test or a change in behaviour due to this work?

Reply:
```
Could this be resolved by providing a date and time if that is what is expected?
```
It is resolved by passing a day in UTC.
The day in UTC is the 31, which then gets correctly converted to the 30th when the Tenant's time zone is applied.

```
Why did this test work / pass with the old code and not the new?
```
Probably because they were fast-forwading the day as mentioned in the other major problem topic (The `withZoneRetainFields()` conversation).

```
Is this a bug in the production code, a bug in the test or a change in behaviour due to this work?
```
I think this is evidence towards the problems and concerns that I described in the document/PDF I gave you.

I think this is a design limitation/flaw that was suppressed by a bug (`withZoneRetainFields()`) and then revealed by a test that was written against a value used to match the buggy behavior..

Reply:
> It is resolved by passing a day in UTC.

I'm sorry, I was trying to move away from that aspect. I don't understand what this means. How can a day be in a timezone :-/

Reply:
> I think this is a design limitation/flaw that was suppressed by a bug (withZoneRetainFields()) and then revealed by a test that was written against a value used to match the buggy behavior..

This was a bug in both the production code and test code that no one has previously identified?

Reply:
```
This was a bug in both the production code and test code that no one has previously identified?
```

The problem was noticed here: UICIRC-565.
The test hid the problem by using a faulty test as I have argued above.

Reply:
The original code passes 2017-07-30 .
The Service that receives this stores this LocalDate in a DateTime variable, resulting in the date being converted into a UTC date 2017-07-30T00:00:00+000Z due to missing time information.
Then the expiration date is calculated, which in this case is an end of the day conversion (because it converts that UTC date into the New York time zone), resulting in 2017-07-30T03:59:59.000Z.
The test then attempts to check to see if the resulting date (from the line above) is 2017-07-31T03:59:59.000Z, which it is not.

The reason the original code passes 2017-07-30 rather than 2017-07-31 is because the local date in the New York time zone for the date in UTC of 2017-07-31T00:00:00+000Z  is the 2017-07-30 and happens to work because the strategy used to determine the end of the day was altering the date and time as mentioned in the conversation we've already had in regards to UICIRC-565.

The solution is to that, if we are passing a date that is expected to be in UTC, then pass a UTC date.
That date in UTC is in fact the 30th because the DateTime always sets the time to 00:00:00 when it stores a LocalDate as a Date.

How this is happening:

This is happening automatically here on this line: https://github.com/folio-org/mod-circulation/blob/440d37bee797dd6ea680f1573168c303117f5a3d/src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java#L76

When the mapping process for Request::from happens, the "Date" string gets interpreted as a "DateTime" string.
Lacking any time, this is where it gets setup as UTC at 00:00:00.

Then the immediately following line changes the timezone here: https://github.com/folio-org/mod-circulation/blob/440d37bee797dd6ea680f1573168c303117f5a3d/src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java#L77

The truncation then improperly calls withZoneRetainFields() rather than calling withZone() here: https://github.com/folio-org/mod-circulation/blob/440d37bee797dd6ea680f1573168c303117f5a3d/src/main/java/org/folio/circulation/domain/Request.java#L352

The reason for changing `withZoneRetainFields()` into `withZone()` is already described in the conversation about UICIRC-565 (as well as in the PR's reasons and learning). It is the same problem, just a different area of code.